aten/src/ATen/InferSize.h

```diff
@@ -37,7 +37,7 @@ inline void infer_size_impl(
     }
   }
 
-  if (numel == newsize || (infer_dim && newsize > 0 && numel % newsize == 0)) {
+  if (TORCH_GUARD_SIZE_OBLIVIOUS(sym_eq(numel, newsize)) || (infer_dim && newsize > 0 && numel % newsize == 0)) {
```

Comment:
Hmm, one question about size-oblivious guards:

So far I've thought their main use case as being for when we have some pytorch code that diverges on two cases (one for the 0/1 case and one for the general case), and we've decided that the general case is safe to use all the time when our size variable is unbacked. The example in my head being that `squeeze()` will repeatedly remove consecutive dims of size 1, which we will ignore when our dims are unbacked.

Here, every single size check in this function is effectively an error check (if the sizes don't match then our reshape should error).

So size oblivious guards don't seem wrong here (the "correct" path is probably not to error). But do you think that `TORCH_SYM_CHECK` is more the right tool for this case? Or maybe I'm holding this wrong.

Reply:
You are right that a deferred runtime assert might be even better. But it's nontrivial to convert this into a `torch._check` because of the disjunction. It may be possible, I just went for the simplest thing in this PR.